VirtoCommerce.Storefront/Controllers/AccountController.cs

```diff
@@ -566,35 +567,76 @@ public async Task<ActionResult> ForgotLogin(ForgotPassword formModel)
             return View("customers/forgot_login", WorkContext);
         }
 
-        [HttpGet("resetpassword")]
+        [HttpPost("forgotpasswordbycode")]
         [AllowAnonymous]
-        public async Task<ActionResult> ResetPassword(string token, string userId)
+        public async Task<ActionResult> ResetPasswordByCode(ResetPasswordByCodeModel formModel)
         {
-            if (string.IsNullOrEmpty(token) || string.IsNullOrEmpty(userId))
+            if (string.IsNullOrEmpty(formModel.Email))
             {
-                WorkContext.ErrorMessage = "Error in URL format";
+                WorkContext.ErrorMessage = "Error in request";
                 return View("error", WorkContext);
             }
 
-            var user = await _signInManager.UserManager.FindByIdAsync(userId);
+            if (string.IsNullOrEmpty(formModel.Code))
+            {
+                WorkContext.ErrorMessage = "Code could not be empty";
+                return View("error", WorkContext);
+            }
+
+            if (!_options.UseSmsForPasswordReset)
+            {
+                WorkContext.ErrorMessage = "Reset password by code is turned off.";
+                return View("error", WorkContext);
+            }
+
+            var user = await _signInManager.UserManager.FindByEmailAsync(formModel.Email);
             if (user == null)
             {
                 WorkContext.ErrorMessage = "User was not found.";
                 return View("error", WorkContext);
             }
 
-            var isValidToken = false;
+            var phoneNumber = await _signInManager.UserManager.GetPhoneNumberAsync(user);
+            var isValidToken = await _signInManager.UserManager.VerifyChangePhoneNumberTokenAsync(user, formModel.Code, phoneNumber);
```

Comment:
It's not a _change phone number_ token, it's a _reset password token_

Reply:
We are using ChangePhoneNumberToken intentionally (6 digits code) as ResetPasswordToken generates token that is good to send by email as a part of link (like http://localhost:2082/Electronics/en-US/account/resetpassword?UserId=x&Token=**CfDJ8H9Z13d0HyxJviWis5gewrrlLU5bCgtuZo8gmfCWpQXflXw2RLFxV4crrbaB05Z3UdSNVnXQpVULvpcEtHmj6ayHfN%2FQaizGMKBVj0Qu054oQfd2Q7EvSnJDHfqKVQ%2BsnhaGl8c9qEOZ0MqpiixDNHPQ2QcLxbxt%2BaLickWNAqx5avPvPFOuuwphTmYpCYJ2AHPEZLda9iM8bJpblDOEDDZMNzJ1UWn%2BgSVhXG9uPBzyfT9fL0DEpmon3qCmPqTxbQ%3D%3D**).
There is no ResetPasswordToken suitable for sms sending in ASP NET Core Indentity.

Reply:
Also please note this is not latest commit for review, as SMS sending part will be rework using Sms notifications implemented on platform level.

Reply:
The problem is that the purpose of this token is to change the phone number:
```csharp
    public virtual Task<string> GenerateChangePhoneNumberTokenAsync(TUser user, string phoneNumber)
    {
      ThrowIfDisposed();
      return GenerateUserTokenAsync(user, Options.Tokens.ChangePhoneNumberTokenProvider, "ChangePhoneNumber:" + phoneNumber);
    }
```

Here is more correct way:
```csharp
var token = await _signInManager.UserManager.GenerateUserTokenAsync(user, TokenOptions.DefaultPhoneProvider, "ResetPassword");
var isValidToken = await _signInManager.UserManager.VerifyUserTokenAsync(user, TokenOptions.DefaultPhoneProvider, "ResetPassword", token);
```

Reply:
Changed, thanks for the help!